threads.go

```diff
@@ -0,0 +1,357 @@
+	if err != nil {
+		return err
+	}
+	err = t.dagService.AddMany(ctx, []format.Node{rec, event, header, body})
+	if err != nil {
+		return err
+	}
+
+	// Update head
+	t.SetHead(settings.Thread, lg.ID, rec.Cid())
+	return nil
+}
+
+func (t *threads) Pull(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, opts ...tserv.PullOption) ([]thread.Record, error) {
+	settings := tserv.PullOptions(opts...)
+
+	if t.PubKey(id, lid) == nil {
+		return nil, fmt.Errorf("could not find log")
+	}
+
+	if t.PrivKey(id, lid) != nil { // This is our own log
+		return t.pullLocal(ctx, id, lid, offset, settings.Limit)
+	}
+
+	// Pull from addresses
+	return t.service.pull(ctx, id, lid, offset, settings)
+}
+
+func (t *threads) Logs(id thread.ID) []thread.LogInfo {
+	lgs := make([]thread.LogInfo, 0)
+	for _, lid := range t.ThreadInfo(id).Logs {
+		lg := t.LogInfo(id, lid)
+		lgs = append(lgs, lg)
+	}
+	return lgs
+}
+
+func (t *threads) Delete(ctx context.Context, id thread.ID) error {
+	panic("implement me")
+}
+
+func (t *threads) getPrivKey() ic.PrivKey {
+	return t.host.Peerstore().PrivKey(t.host.ID())
+}
+
+func (t *threads) createLog() (info thread.LogInfo, err error) {
+	return util.CreateLog(t.host.ID())
+}
+
+func (t *threads) getOrCreateLog(id thread.ID, lid peer.ID) (info thread.LogInfo, err error) {
+	info = t.LogInfo(id, lid)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+func (t *threads) getOwnLog(id thread.ID) (info thread.LogInfo) {
+	for _, lid := range t.LogsWithKeys(id) {
+		if t.PrivKey(id, lid) != nil {
+			info = t.LogInfo(id, lid)
+			return
+		}
+	}
+	return
+}
+
+func (t *threads) getOrCreateOwnLog(id thread.ID) (info thread.LogInfo, err error) {
+	info = t.getOwnLog(id)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+func (t *threads) createNode(ctx context.Context, body format.Node, lg thread.LogInfo, settings *tserv.AddSettings) (thread.Record, error) {
+	if settings.Key == nil {
+		var key []byte
+		logKey := t.ReadKey(settings.Thread, settings.KeyLog)
+		if logKey != nil {
+			key = logKey
+		} else {
+			key = lg.ReadKey
+		}
+		var err error
+		settings.Key, err = crypto.ParseEncryptionKey(key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	event, err := cbor.NewEvent(ctx, t.dagService, body, settings)
+	if err != nil {
+		return nil, err
+	}
+
+	prev := cid.Undef
+	if len(lg.Heads) != 0 {
+		prev = lg.Heads[0]
+	}
+	fk, err := crypto.ParseEncryptionKey(lg.FollowKey)
+	if err != nil {
+		return nil, err
+	}
+	rec, err := cbor.NewRecord(ctx, t.dagService, event, prev, lg.PrivKey, fk)
+	if err != nil {
+		return nil, err
+	}
+
+	t.SetHead(settings.Thread, lg.ID, rec.Cid())
+
+	return rec, nil
+}
+
+func (t *threads) pullLocal(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, limit int) ([]thread.Record, error) {
```

Comment:
This looks for new records ahead of the offset _locally_. This is called locally when requesting from a log addressed to this peer, and remotely via the RPC service when requesting as a remote client.